x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java

```diff
@@ -73,12 +78,60 @@ void setMetadata(Metadata metadata) {
+     * to another thread. Under contention the searchDispatcher is only called once when the value is not in the cache. The
+     * searchDispatcher should schedule the search / callback _asynchronously_ because if the searchDispatcher blocks, then this
+     * method will block. The callback is call on the thread calling this method or under cache miss and contention, the thread running
+     * the part of the searchDispatcher that calls the callback.
+     * @param searchRequest the cache key and input for the search dispatcher
+     * @param searchDispatcher the logical block to be called on cache miss
+     * @param callBack the callback which gets the value asynchronously, which could be a searchResponse or exception (negative lookup)
+     */
+    public void resolveOrDispatchSearch(
+        SearchRequest searchRequest,
+        BiConsumer<SearchRequest, ActionListener<SearchResponse>> searchDispatcher,
+        BiConsumer<SearchResponse, Exception> callBack
+    ) {
+        String enrichIndex = getEnrichIndexKey(searchRequest);
+        CacheKey cacheKey = new CacheKey(enrichIndex, searchRequest);
+        try {
+            CompletableFuture<SearchResponse> cacheEntry = cache.computeIfAbsent(cacheKey, request -> {
+                CompletableFuture<SearchResponse> completableFuture = new CompletableFuture<>();
+                searchDispatcher.accept(searchRequest, wrap(completableFuture::complete, completableFuture::completeExceptionally));
+                return completableFuture;
+            });
+            cacheEntry.whenComplete((response, throwable) -> {
+                if (throwable != null) {
+                    scheduleClearFailure(cacheKey, cacheEntry);
+                    if (throwable instanceof Exception) {
+                        callBack.accept(response, (Exception) throwable);
+                        return;
+                    } else {
+                        // Let ElasticsearchUncaughtExceptionHandler handle this, should it deem it non-fatal let's still trigger the
+                        // callback to ensure proper clean up.
+                        callBack.accept(response, new Exception("Throwable was raised '" + throwable.getMessage() + "'"));
+                        throw (Error) throwable;
+                    }
+                }
+                callBack.accept(response, null);
+            });
+        } catch (ExecutionException e) {
+            callBack.accept(null, e);
+        }
+    }
+
     private String getEnrichIndexKey(SearchRequest searchRequest) {
         String alias = searchRequest.indices()[0];
         IndexAbstraction ia = metadata.getIndicesLookup().get(alias);
         return ia.getIndices().get(0).getIndex().getName();
     }
 
+    private void scheduleClearFailure(CacheKey cacheKey, CompletableFuture<SearchResponse> cacheEntry) {
+        // For now, we schedule immediately. But perhaps we find a reason to cache failures as well.
```

Comment:
If an error occurs searching an enrich index then I don't think this should be cached. The search requests are controlled by the processor / policy type and not by the user. So an exception should be seen as a bug.

Reply:
I've removed the premature indirection and inlined the invalidation. https://github.com/elastic/elasticsearch/pull/77259/commits/c38f0970ecc6eeadca47f52162eb3aa6263c06c5